### ISSUE TITLE
How do you just provide some credentials?

### ISSUE THREAD

**Issue by camertron** [google/google-auth-library-ruby#75]
We are trying to upgrade to the newest google-api-client library (v0.9.9) and are struggling with getting the authentication to work. How do we simply provide a client_id, client_secret, access_token, and refresh_token? We tried using Signet directly but that doesn't seem to work.

```ruby
client.authorization = Signet::OAuth2::Client.new(
  client_id: "...",
  client_secret: "...",
  access_token: "...",
  refresh_token: "..."
)
```
We then tried to use this library, but there doesn't seem to be a straightforward way to provide credentials. Any help would be appreciated!

**Comment by camertron** [comment]
I also tried this:

```ruby
::Google::Apis::AndroidpublisherV2::AndroidPublisherService.new.tap do |client|
    secrets = ::Google::APIClient::ClientSecrets.new(
      flow: :web,
      web: {
        client_id:     '...',
        client_secret: '...',
        access_token:  '...',
        refresh_token: '...'
      }
    )

    client.authorization = secrets.to_authorization
  end
end
```
Unfortunately I always get an `Unauthorized` error. These are the same credentials that worked (and are still working) in production for us with v0.8 of the google-api-client gem. Any ideas?

**Comment by camertron** [comment]
I did a bunch of additional digging and discovered that v0.8.6 works because it automatically refreshes stale access tokens. I discovered this by setting the logger:

```ruby
require 'logger'
Google::APIClient.logger = Logger.new(STDOUT)  # for v0.8.6
Google::Apis.logger = Logger.new(STDOUT)  # for v0.9.9
```
Running my test script against the 0.8.6 version prints a 401 response to the original API call, then another request with a refreshed access token that results in a 200 successful response. The 0.9.9 version on the other hand makes the exact same request twice, without attempting to refresh the access token (you can see the access token in the request header).

This is strange behavior, considering that it appears this [monkeypatch](https://github.com/google/google-auth-library-ruby/blob/master/lib/googleauth/signet.rb#L42) to Signet is supposed to handle token refreshing. The monkeypatch adds a method called `#apply!` to `Signet::OAuth2::Client`, which google-api-client [checks for](https://github.com/google/google-api-ruby-client/blob/master/lib/google/apis/core/http_command.rb#L137) in order to determine if the chosen authorization client is refreshable. So, if tokens are supposed to be automatically refreshed, what's going on?

v0.8.6 of the google-api-client library didn't concern itself with `#apply!` and called a method named `#fetch_access_token!` [directly on the Signet client](https://github.com/google/google-api-ruby-client/blob/v0.8.6/lib/google/api_client.rb#L728) whenever an API call returned a 401. In v0.9.9, `#apply!` wraps the `#fetch_access_token!` method and only calls it if `@access_token` is `nil` (or the token has expired). However, nowhere that I've been able to find is`@access_token` ever set to `nil` on a 401, meaning the access token will never get refreshed.

One potential solution is to skip the `nil` check [here](https://github.com/google/google-auth-library-ruby/blob/master/lib/googleauth/signet.rb#L45). Here's a monkeypatch:

```ruby
module Signet
  module OAuth2
    class Client
      def apply!(a_hash, opts = {})
        fetch_access_token!(opts)
        a_hash[AUTH_METADATA_KEY] = "Bearer #{access_token}"
      end
    end
  end
end
```
This works, but I don't think it's the right answer. Skipping the `nil` check means we'll be requesting a new access token on every request.

Another potential solution might be to modify the [`refresh_authorization` method](https://github.com/google/google-api-ruby-client/blob/master/lib/google/apis/core/http_command.rb#L129) in google-api-client to either call `#fetch_access_token!` or set `@access_token` to `nil` somehow. However, it looks like you can configure google-api-client to use a number of different kinds of authorization client - they just have to conform to some common interface. There would have to be a method added to that interface that handles token refreshes.

I'm not sure what the solution would be - what do the maintainers think?

**Comment by sqrrrl** [comment]
I think the issue is `expires_within?` is broken when there's no time info on the token. Currently it's

```ruby
return self.expires_at != nil && Time.now >= (self.expires_at - sec)
```

which returns true if expires_at is nil...should be:

```ruby
return self.expires_at.nil? || Time.now >= (self.expires_at - sec)
```

I'll fix it in signet on Monday. Workaround is either set the correct time or just mark as expired (below):

```ruby
client.authorization.expires_in = 0
```

**Comment by camertron** [comment]
@sqrrrl awesome, that totally makes sense. I didn't even consider the expiration case :) If I set `expires_in` as you suggest:

```ruby
client.authorization.expires_in = 0
```
will that cause an access token refresh on every API request?

**Comment by sqrrrl** [comment]
Nope. It should correctly get the expiration time on the next refresh and will be valid so long as you continue to use that credentials instance. The problem is introduced when it's serialized/deserialized and the expiration time is lost.

**Comment by camertron** [comment]
@sqrrrl wonderful, thanks!

**Comment by sqrrrl** [comment]
Fixed in signet 0.7.3. Just refresh your dependencies and you should be OK now.
